atc/db/resource.go

```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+		WHERE i.resource_id!=$1
+)
+SELECT * FROM build_ids
+`
+)
+
+// getCausalityBuilds figures out all the builds that are related to a particular resource version
+// This can include builds that were used the resource version (and its descendents) as an input,
+// and builds that generated some ancestor of the build that generated the resource version itself.
+func (r *resource) getCausalityBuilds(versionMD5 string, query string) ([]int, error) {
+	buildIDs := make([]int, 0)
+
+	// downstream builds that were caused by this resource version
+	rows, err := r.conn.Query(query, r.id, versionMD5)
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		var buildID int
+		err := rows.Scan(&buildID)
+		if err != nil {
+			return nil, err
+		}
+		buildIDs = append(buildIDs, buildID)
+	}
+
+	return buildIDs, nil
+}
+
+// this allows us to reuse getCausalityResourceVersions to construct both upstream and downstream trees by passing in a different updater fn
+type resourceVersionUpdater func(*atc.CausalityResourceVersion, *atc.CausalityBuild)
+
+// for a given list of build ids, use the build inputs and outputs to construct the tree.
+// the handleInput and handleOutput functions can be used to reverse the direction of the tree:
+// i.e. if the children of a build node is the output, then it's a downstream tree, vice versa if it's an input
+func (r *resource) getCausalityResourceVersions(
+	buildIDs []int, root *atc.CausalityResourceVersion,
+	handleInput resourceVersionUpdater,
+	handleOutput resourceVersionUpdater,
+) error {
+	// construct the job and build nodes. These are placed into a map for easy access down the line
+	rows, err := psql.Select("b.id", "b.name", "b.status", "j.id", "j.name").
+		From("builds b").
+		Join("jobs j ON b.job_id = j.id").
+		Where(sq.Eq{"b.id": buildIDs}).
+		RunWith(r.conn).
+		Query()
+	if err != nil {
+		return err
+	}
+
+	builds := make(map[int]*atc.CausalityBuild)
+	for rows.Next() {
+		var buildID, jobID int
+		var buildName, jobName, status string
+
+		rows.Scan(&buildID, &buildName, &status, &jobID, &jobName)
+
+		if _, found := builds[buildID]; !found {
+			builds[buildID] = &atc.CausalityBuild{
+				ID:      buildID,
+				JobID:   jobID,
+				Name:    buildName,
+				JobName: jobName,
+				Status:  atc.BuildStatus(status),
+			}
+		}
+	}
+
+	resourceVersions := make(map[int]*atc.CausalityResourceVersion)
+	// pre-populate the list with the root
+	resourceVersions[root.ResourceVersionID] = root
+
+	// go through all the inputs and construct the struct, this will partially construct the tree
+	rows, err = psql.Select("r.id", "rcv.id", "r.name", "rcv.version", "i.build_id").
+		From("build_resource_config_version_inputs i").
+		Join("resources r ON r.id = i.resource_id").
+		Join("resource_config_versions rcv ON rcv.version_md5 = i.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id").
+		Where(sq.Eq{"i.build_id": buildIDs}).
```

Comment:
Is it feasible to do this without a `WHERE build_id IN (....)` condition? I imagine that the number of build ids could be pretty huge, so doing a `JOIN` would probably be more scalable.

e.g.

```sql
WITH RECURSIVE build_ids AS
(
...
)
SELECT ...
FROM build_resource_config_version_inputs i
JOIN ...
JOIN build_ids bi ON i.build_id = bi.id
...
UNION ALL
SELECT ...
FROM build_resource_config_version_outputs o
JOIN ...
JOIN build_ids bi ON o.build_id = bi.id
...
```

Reply:
You mean by embedding the `WITH RECURSIVE` query inside each of the queries for builds, build inputs, and build outputs? Is postgres capable of "learning" from the result of previous executions and make more informed plans?

My main concern is that the `WITH RECURSIVE` query should be the most expensive one, so we should only run it once and use the results of that in the 3 cheaper queries.

Reply:
Or is it possible to do this in a transaction, store the result of the `WITH RECURSIVE` as a temp view/table, and then refer to that table in subsequent queries in the same transaction?

Reply:
Well at least the inputs and outputs can be a single query with the `UNION ALL`, but yeah I was forgetting about the "build info" query

I know that the list passed to a `WHERE ... IN (...)`  has a limit of 65000 or so elements - but at that point, the performance is already probably pretty bad. Temp table is an option but feels pretty heavy weight, especially for the common case which shouldn't be too crazy.

Tagging @clarafu - WDYT?

Reply:
I just want to point out that it needs to execute different functions depending on if it's processing the inputs or outputs. Which should be easy to do by adding a `'input' AS type` to the select statement and then special casing on that in the code

https://github.com/concourse/concourse/blob/a7b862ea40fa15d308ff383784b46d9f5fa58345/atc/db/resource.go#L1176-L1183

Reply:
re: https://github.com/concourse/concourse/pull/7125#issuecomment-855015082:

It looks like the recursive query is actually quite fast

```
CTE Scan on build_ids  (cost=694.57..709.99 rows=771 width=8) (actual time=1.119..196.107 rows=17056 loops=1)
  CTE build_ids
    ->  Recursive Union  (cost=8.59..694.57 rows=771 width=8) (actual time=1.118..191.759 rows=17056 loops=1)
          ->  Unique  (cost=8.59..8.59 rows=1 width=8) (actual time=1.117..1.278 rows=661 loops=1)
                ->  Sort  (cost=8.59..8.59 rows=1 width=8) (actual time=1.116..1.173 rows=661 loops=1)
                      Sort Key: i.build_id
                      Sort Method: quicksort  Memory: 55kB
                      ->  Index Scan using build_inputs_resource_versions_idx on build_resource_config_version_inputs i  (cost=0.56..8.58 rows=1 width=8) (actual time=0.022..0.973 rows=661 loops=1)
                            Index Cond: ((resource_id = 101) AND (version_md5 = '40fa80af72b9c40a7b218fd3fb5393c0'::text))
          ->  HashAggregate  (cost=66.29..67.06 rows=77 width=8) (actual time=26.320..26.764 rows=2342 loops=7)
                Group Key: bp.to_build_id
                ->  Nested Loop  (cost=0.98..66.09 rows=77 width=8) (actual time=0.566..22.632 rows=21790 loops=7)
                      Join Filter: (bi.build_id = i_1.build_id)
                      ->  Nested Loop  (cost=0.42..48.73 rows=14 width=24) (actual time=0.557..6.702 rows=2535 loops=7)
                            ->  WorkTable Scan on build_ids bi  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.463 rows=2437 loops=7)
                            ->  Index Only Scan using build_pipes_from_build_id_to_build_id_uniq on build_pipes bp  (cost=0.42..4.84 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=17056)
                                  Index Cond: (from_build_id = bi.build_id)
                                  Heap Fetches: 663
                      ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i_1  (cost=0.56..1.17 rows=6 width=8) (actual time=0.003..0.004 rows=9 loops=17745)
                            Index Cond: (build_id = bp.from_build_id)
                            Filter: (resource_id <> 101)
                            Rows Removed by Filter: 0
                            Heap Fetches: 12926
Planning time: 0.469 ms
Execution time: 197.186 ms
```

On the other hand, the massing `WHERE build_id IN (...)` condition seems to be really slow - the `EXPLAIN ANALYZE` has been running for over 3 minutes at this point, still not complete.

So, doing it as a `JOIN` seems like a better option, even if it means needing to run the query twice. For this query:

```sql
WITH RECURSIVE build_ids AS (
		SELECT DISTINCT i.build_id
			FROM build_resource_config_version_inputs i
			WHERE i.resource_id=101 AND i.version_md5='40fa80af72b9c40a7b218fd3fb5393c0'
	UNION ALL
		SELECT DISTINCT bp.to_build_id AS build_id
		FROM build_ids bi
		INNER JOIN build_pipes bp ON bi.build_id = bp.from_build_id
		INNER JOIN build_resource_config_version_inputs i ON i.build_id = bi.build_id
		WHERE i.resource_id!=101
)
SELECT r.id, rcv.id, r.name, rcv.version, i.build_id, 'input' as type
FROM build_resource_config_version_inputs i
JOIN resources r ON r.id = i.resource_id
JOIN resource_config_versions rcv ON rcv.version_md5 = i.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id
JOIN build_ids bi ON i.build_id = bi.build_id
UNION ALL
SELECT r.id, rcv.id, r.name, rcv.version, o.build_id, 'output' as type
FROM build_resource_config_version_outputs o
JOIN resources r ON r.id = o.resource_id
JOIN resource_config_versions rcv ON rcv.version_md5 = o.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id
JOIN build_ids bi ON o.build_id = bi.build_id
```

...here's the `EXPLAIN ANALYZE`:

```
Append  (cost=1535.45..54855.85 rows=3 width=113) (actual time=3.821..1212.134 rows=134774 loops=1)
  CTE build_ids
    ->  Recursive Union  (cost=8.59..694.57 rows=771 width=8) (actual time=1.654..215.981 rows=17090 loops=1)
          ->  Unique  (cost=8.59..8.59 rows=1 width=8) (actual time=1.653..1.983 rows=661 loops=1)
                ->  Sort  (cost=8.59..8.59 rows=1 width=8) (actual time=1.652..1.826 rows=661 loops=1)
                      Sort Key: i_1.build_id
                      Sort Method: quicksort  Memory: 55kB
                      ->  Index Scan using build_inputs_resource_versions_idx on build_resource_config_version_inputs i_1  (cost=0.56..8.58 rows=1 width=8) (actual time=0.026..1.472 rows=661 loops=1)
                            Index Cond: ((resource_id = 101) AND (version_md5 = '40fa80af72b9c40a7b218fd3fb5393c0'::text))
          ->  HashAggregate  (cost=66.29..67.06 rows=77 width=8) (actual time=28.771..29.864 rows=2347 loops=7)
                Group Key: bp.to_build_id
                ->  Nested Loop  (cost=0.98..66.09 rows=77 width=8) (actual time=0.472..24.682 rows=21834 loops=7)
                      Join Filter: (bi_2.build_id = i_2.build_id)
                      ->  Nested Loop  (cost=0.42..48.73 rows=14 width=24) (actual time=0.462..7.386 rows=2540 loops=7)
                            ->  WorkTable Scan on build_ids bi_2  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.441 rows=2441 loops=7)
                            ->  Index Only Scan using build_pipes_from_build_id_to_build_id_uniq on build_pipes bp  (cost=0.42..4.84 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=17090)
                                  Index Cond: (from_build_id = bi_2.build_id)
                                  Heap Fetches: 697
                      ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i_2  (cost=0.56..1.17 rows=6 width=8) (actual time=0.003..0.005 rows=9 loops=17779)
                            Index Cond: (build_id = bp.from_build_id)
                            Filter: (resource_id <> 101)
                            Rows Removed by Filter: 0
                            Heap Fetches: 13250
  ->  Nested Loop  (cost=840.88..39800.98 rows=2 width=113) (actual time=3.821..1111.347 rows=127307 loops=1)
        ->  Hash Join  (cost=840.33..4550.73 rows=4254 width=63) (actual time=3.704..432.316 rows=127835 loops=1)
              Hash Cond: (i.resource_id = r.id)
              ->  Nested Loop  (cost=0.56..3652.47 rows=4254 width=45) (actual time=1.679..379.326 rows=127835 loops=1)
                    ->  CTE Scan on build_ids bi  (cost=0.00..15.42 rows=771 width=8) (actual time=1.656..221.465 rows=17090 loops=1)
                    ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i  (cost=0.56..4.66 rows=6 width=45) (actual time=0.006..0.008 rows=7 loops=17090)
                          Index Cond: (build_id = bi.build_id)
                          Heap Fetches: 11169
              ->  Hash  (cost=807.12..807.12 rows=2612 width=22) (actual time=1.983..1.983 rows=2612 loops=1)
                    Buckets: 4096  Batches: 1  Memory Usage: 174kB
                    ->  Seq Scan on resources r  (cost=0.00..807.12 rows=2612 width=22) (actual time=0.011..1.488 rows=2612 loops=1)
        ->  Index Scan using resource_config_scope_id_and_version_md5_unique on resource_config_versions rcv  (cost=0.56..8.28 rows=1 width=92) (actual time=0.005..0.005 rows=1 loops=127835)
              Index Cond: ((resource_config_scope_id = r.resource_config_scope_id) AND (version_md5 = i.version_md5))
  ->  Nested Loop  (cost=1.25..14360.26 rows=1 width=113) (actual time=0.048..85.249 rows=7467 loops=1)
        ->  Nested Loop  (cost=0.70..3024.50 rows=1368 width=63) (actual time=0.032..50.585 rows=8122 loops=1)
              ->  Nested Loop  (cost=0.42..2545.01 rows=1368 width=45) (actual time=0.019..36.930 rows=8122 loops=1)
                    ->  CTE Scan on build_ids bi_1  (cost=0.00..15.42 rows=771 width=8) (actual time=0.001..3.806 rows=17090 loops=1)
                    ->  Index Only Scan using build_resource_config_version_outputs_uniq on build_resource_config_version_outputs o  (cost=0.42..3.26 rows=2 width=45) (actual time=0.002..0.002 rows=0 loops=17090)
                          Index Cond: (build_id = bi_1.build_id)
                          Heap Fetches: 185
              ->  Index Scan using resources_pkey on resources r_1  (cost=0.28..0.34 rows=1 width=22) (actual time=0.001..0.001 rows=1 loops=8122)
                    Index Cond: (id = o.resource_id)
        ->  Index Scan using resource_config_scope_id_and_version_md5_unique on resource_config_versions rcv_1  (cost=0.56..8.28 rows=1 width=92) (actual time=0.004..0.004 rows=1 loops=8122)
              Index Cond: ((resource_config_scope_id = r_1.resource_config_scope_id) AND (version_md5 = o.version_md5))
Planning time: 3.089 ms
Execution time: 1220.851 ms
```

Reply:
Hmm ya I definitely get the concern but right now I can't think of an easy way around it. 🤔  I was also thinking iof we can just combine the queries so we don't have to pass around the list of builds but I don't think we can attach fetching the resource/version/build info in the recursive query. The parent query that runs in order to do the `UNION ALL` will be super slow if it tries to union on all the resource/version/build info rather than just doing the union on the builds and then we fetch out the info after. 

As an example, I tried running this against `hush-house` with approx 200 builds and it took around 800 ms to run. https://explain.depesz.com/s/8rfJ

I think if we can think of a better way to do it that is easy enough then yes we should probably go ahead with that , but for now I think this current approach is good enough for this first iteration!

Reply:
@clarafu thoughts on just running the recursive query 2 (or 3) times, per https://github.com/concourse/concourse/pull/7125#discussion_r646657959? That last `EXPLAIN ANALYZE` is run on a resource version with 17000 builds and it doesn't get *too* slow